server/pulp/server/async/tasks.py

```diff
@@ -255,7 +255,7 @@ def apply_async_with_reservation(self, resource_type, resource_id, *args, **kwar
         # To avoid the race condition where __call__ method below is called before
         # this change is propagated to all db nodes, using an 'upsert' here and setting
         # the task state to 'waiting' only on an insert.
-        task_status.save(fields_to_set_on_insert=['state', 'start_time'])
+        task_status.save(fields_to_set_on_insert=['state', 'start_time'], safe=True)
```

Comment:
On 05/22/2015 09:03 AM, bmbouter wrote:
> IMHO, this is a bug in mongoDB because the replica set is configured for
> one level of safety, but a client can override it. That doesn't seem
> right.

I think it is useful that we can specify a write-concern on any
particular query. There are times that data isn't particularly important
to record so safely, and the write concern makes the query take longer
to return. For example, we record all of the heartbeats from the workers
in the database. I don't think it's critical that those writes fsync or
propagate, because it'll be ok if the next one is the one that makes it.
I think we would rather just hand it to mongo and hope for the best in
that situation. Does that make sense? It's up to the application writer
to know when we can override the default write concern in either
direction (safer or less safe).

In the Mongo work I've done in the past with large databases, I took
great advantage of being able to specify which queries were critically
important, and which ones were OK to miss when eventual consistency is OK.

> I remember
> you mentioning that they brought back the |w:majority| setting in 3.0,
> which keeps the client out of the business of specifying the replication
> level, but we still use mongoDB 2.x.

We should read more about this. It does say that something about this
changed in 3.0, and maybe also in 2.6? The wording confused me, but I
think I might have been wrong that majority was new in 3.0. I think
w:majority might actually exist in some 2.x versions, but maybe it means
different things or something? We could use the lab to test this by
switching out the Vagrant box OS to get differing versions of Mongo.

> Again, great investigation work.

Thanks!

-- 
Randy Barlow